Preliminary support for Matter cover

hidaris opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds matter cover control feature. Curtain tilt support has not been added yet.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant-libs/python-matter-server/issues/225
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/26857

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

hidaris committed:
preliminary support for Matter cover

MartinHjelmare commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,157 @@
+
+        assert current_position is not None
+
+        LOGGER.info(
```
Please don't do side effects like logging in entity state properties. We can remove this.

MartinHjelmare commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,157 @@
+    @property
+    def supported_features(self) -> CoverEntityFeature:
+        """Flag supported features."""
+        features = (
```
If it's static we can use a class attribute, `_attr_supported_features`.

MartinHjelmare commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,157 @@
+        LOGGER.debug(
+            "GOT OPERATIONAL STATUS %s for %s", operational_status, self.entity_id
+        )
+        state = operational_status & 0b11
```
Please use constants with descriptive names for these integer values.

marcelveldt replied:
> 


> Please use integers instead.

This is more easy to recognize when reading the Matter specs so I'd vote to leave it like this actually
or indeed put it in a constant with a comment behind it with the hex value

MartinHjelmare replied:
Ok. A constant with the value is ok.

MartinHjelmare commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,157 @@
+        self._attr_current_cover_position = self.get_matter_attribute_value(
+            clusters.WindowCovering.Attributes.CurrentPositionLiftPercentage
+        )
+        LOGGER.info("GOT CURRENT POSITION %s", self._attr_current_cover_position)
```
```suggestion
        LOGGER.debug("Current position: %s", self._attr_current_cover_position)
```

MartinHjelmare commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,157 @@
+            clusters.WindowCovering.Attributes.OperationalStatus,
+        ),
+        optional_attributes=(),
+        # restrict device type to prevent discovery in switch platform
```
This is not what happens. It stops on off plug in unit device attributes from getting discovered as covers.

hidaris replied:
I don't particularly understand the way this is handled, the matter cluster can get a definite device type, can't we use that information?

marcelveldt replied:
You can but you have overlapping device types. This was just something you can ignore here, I believe you copied it from another platform ?

hidaris replied:
I may need to familiarize myself with the implementation again.

MartinHjelmare reviewed (changes requested):
Please also add tests. See the other platforms for examples of tests using the diagnostic dumps stored as fixtures.

marcelveldt commented:
Hi @hidaris do you have time/interest to address the feedback comments or may/can I help you finishing it ?

hidaris commented:
> Hi @hidaris do you have time/interest to address the feedback comments or may/can I help you finishing it ?

Most of the feedbacks I've fixed locally, but because of the network in my region, the tests I've added don't have environmental testing at the moment, and I'm working on that...

hidaris committed:
add matter cover test, tweak some code structure

hidaris commented:
I added the test, but it keeps telling me I don't have the entity id, I used flag --snapshot-update, but it didn't solve the problem, can anyone tell me what's wrong?

hidaris committed:
rename matter cover test case

MartinHjelmare commented on tests/components/matter/test_cover.py:
```diff
@@ -0,0 +1,126 @@
+
+
+@pytest.fixture(name="window_covering")
+async def door_lock_fixture(
```
Please update the name.

marcelveldt commented on homeassistant/components/matter/cover.py:
```diff
@@ -23,31 +24,39 @@
+class OperationalStatus(IntFlag):
+    """Currently ongoing operations enumeration for coverings, as defined in the Matter spec."""
+
+    coveringIsCurrentlyNotMoving = 0b00
```
this is how it is named in the spec but I suggest to change it to python snake case

MartinHjelmare commented:
If you address the lint issues we can run CI again and see the error message of the pytest tests, and figure out what's wrong. I'm guessing the entity_id may be incorrect.

hidaris committed:
change operational status to snake case, reorder match case

hidaris commented:
> If you address the lint issues, we can rerun CI and see the pytest error message and figure out what is wrong. I suspect that the entity_id might be wrong.

I just pushed the fix

MartinHjelmare commented:
Please also address the lint issues: ruff, isort and pylint
Click on the jobs to view the logs and what's wrong.

hidaris committed:
fix lint problem in matter cover

hidaris commented:
> Please also address the lint issues: ruff, isort and pylint Click on the jobs to view the logs and what's wrong.


> Please also address the lint issues: ruff, isort and pylint Click on the jobs to view the logs and what's wrong.

I adjusted the ruff and isort issue, pylint says I have a problem with intflag, I'm not sure if that's a problem

hidaris commented:
Additionally, I receive a prompt indicating that my branch is out-of-date with the base branch. Should I update it using rebase?

MartinHjelmare commented:
You don't need to rebase the branch unless there are merge conflicts.

You do need to resolve the pylint issues though.

hidaris committed:
change operational status from int flag to enum

hidaris commented:
The test error of ci says that the corresponding entity_id cannot be found, I want to know where this id should be found, I see it in the door lock fixture
![image](https://user-images.githubusercontent.com/8336146/228849879-74e68c60-4830-4f98-98ff-4766beb4a253.png)
Is it related to here, local tweaks don't seem to work

MartinHjelmare commented:
The entity_id seems to be `cover.longan_link_wncv_da01`. It's logged if you read the CI log.

hidaris committed:
fix cover entity id in tests

hidaris committed:
fix matter cover tests

hidaris committed:
keep discovery schemas sorted

MartinHjelmare commented:
Can't we set the operational status attribute too in the test so we can test all possible states?

hidaris committed:
recover matter cover state tests

hidaris commented:
> Can't we set the operational status attribute too in the test so we can test all possible states?

Yes, I added it

hidaris commented:
CI says docs-missing, does this need to be fixed?

MartinHjelmare commented:
Yes, please link a docs PR with an addition of the new platform to the `ha_platforms` front matter at the top of the page.

hidaris commented:
> Yes, please link a docs PR with an addition of the new platform to the `ha_platforms` front matter at the top of the page.
> 
> https://www.home-assistant.io/integrations/matter/

I just updated the documentation https://github.com/home-assistant/home-assistant.io/pull/26857

marcelveldt commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,154 @@
+            clusters.WindowCovering.Attributes.CurrentPositionLiftPercentage,
+            clusters.WindowCovering.Attributes.OperationalStatus,
+        ),
+        optional_attributes=(),
```
you can leave this one out

marcelveldt replied:
```suggestion
```

marcelveldt commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,154 @@
+class OperationalStatus(Enum):
+    """Currently ongoing operations enumeration for coverings, as defined in the Matter spec."""
+
+    covering_is_currently_not_moving = 0b00
```
Shouldn't enum keys always be uppercase (like constants) ?

marcelveldt commented on homeassistant/components/matter/cover.py:
```diff
@@ -0,0 +1,154 @@
+
+        state = operational_status & OPERATIONAL_STATUS_MASK
+        match state:
+            case OperationalStatus.covering_is_currently_opening.value:
```
the point of the enum is a bit gone now this way.
suggestions:

1) Add a missing method in the enum to fallback to "unknown" key --> read the operational status as enum.
2) drop the enum and just use constants instead
3) switch to IntEnum so you can do the comparing without the need for the ".value"

hidaris committed:
Adjust the coding style related to OperationalStatus and remove unused code

MartinHjelmare reviewed (approved):
Looks good!

marcelveldt reviewed (approved):
Awesome, thanks for your contribution !

marcelveldt commented:
> @marcelveldt Just wondering if you're working on the Thermostat cluster? I've got some free time next week and was thinking of giving it a shot.

Yes, I am. I'll drop a PR for that this week.
